File: frontend/src/engine/notation.ts
// Algebraic notation — kept in lockstep with backend/app/engine/notation.py.
// Pawn move:   "<col><rank>"        e.g. "e2"
// Wall move:   "<col><rank><h|v>"   e.g. "e3v"
// col 'a'..'i' = engine col 0..8.   rank '1'..'9' = engine row 8..0.
// Walls live on engine row 0..7, so wall ranks are 2..9.

import type { Move } from './gameTypes';

const PAWN_RE = /^([a-i])([1-9])$/;
const WALL_RE = /^([a-i])([2-9])([hv])$/;

export class NotationError extends Error {}

export function parseMove(text: string): Move {
  const s = text.trim().toLowerCase();
  const wall = WALL_RE.exec(s);
  if (wall) {
    const [, colLetter, rankDigit, orient] = wall;
    return {
      kind: 'wall',
      wall: {
        col: colLetter.charCodeAt(0) - 97,
        row: 9 - Number.parseInt(rankDigit, 10),
        orientation: orient as 'h' | 'v',
      },
    };
  }
  const pawn = PAWN_RE.exec(s);
  if (pawn) {
    const [, colLetter, rankDigit] = pawn;
    return {
      kind: 'pawn',
      to: {
        col: colLetter.charCodeAt(0) - 97,
        row: 9 - Number.parseInt(rankDigit, 10),
      },
    };
  }
  throw new NotationError(`unrecognized move notation: ${JSON.stringify(text)}`);
}

export function serializeMove(move: Move): string {
  const colLetter = (c: number) => String.fromCharCode(97 + c);
  if (move.kind === 'pawn') {
    return `${colLetter(move.to.col)}${9 - move.to.row}`;
  }
  return `${colLetter(move.wall.col)}${9 - move.wall.row}${move.wall.orientation}`;
}

File: backend/app/engine/notation.py
from __future__ import annotations

import re

from app.engine.game_types import Move, PawnMove, Position, Wall, WallMove

_PAWN_RE = re.compile(r"^([a-i])([1-9])$")
_WALL_RE = re.compile(r"^([a-i])([2-9])([hv])$")


class NotationError(ValueError):
    """Raised when a move-history string can't be parsed."""


def parse_move(text: str) -> Move:
    s = text.strip().lower()
    m = _WALL_RE.match(s)
    if m:
        col_letter, rank_digit, orient = m.groups()
        col = ord(col_letter) - 97
        row = 9 - int(rank_digit)
        return WallMove(wall=Wall(row=row, col=col, orientation=orient))  # type: ignore[arg-type]
    m = _PAWN_RE.match(s)
    if m:
        col_letter, rank_digit = m.groups()
        col = ord(col_letter) - 97
        row = 9 - int(rank_digit)
        return PawnMove(to=Position(row=row, col=col))
    raise NotationError(f"unrecognized move notation: {text!r}")


def serialize_move(move: Move) -> str:
    if isinstance(move, PawnMove):
        return f"{chr(97 + move.to.col)}{9 - move.to.row}"
    return f"{chr(97 + move.wall.col)}{9 - move.wall.row}{move.wall.orientation}"
